File: src/rev1/test/example_reader.f90
! This is an example of using the avmesh library with Fortran90.
! Error checking is ignored for brevity.
program example
   implicit none
   integer :: avmid, meshCount
   real*8 :: modelScale, ref_point(3)
   character*128 :: description, contactInfo
   character*128 :: meshName, meshType
   integer :: jmax,kmax,lmax
   integer :: nBndEdges,nPtsPerStrand
   integer :: nNodes,nEdges,nFaces
   integer :: nLevels,nBlocks,nFringe
   integer :: i, istat

   call avm_read_headersf(avmid, 'mixed.avm', istat)
   if (istat.ne.0) stop 'ERROR: not an AVMesh file'

   call avm_get_intf(avmid, 'meshCount', meshCount, istat)
   call avm_get_realf(avmid, 'modelScale', modelScale, istat)
   call avm_get_strf(avmid, 'contactInfo', contactInfo, istat)
   call avm_get_real_arrayf(avmid, 'referencePoint', ref_point, 3, istat)

   print '(A,1X,I0)', 'Mesh Count is', meshCount

   do i=1,meshCount
      call avm_selectf(avmid, 'mesh', i, istat)
      call avm_get_strf(avmid, 'meshName', meshName, istat)
      call avm_get_strf(avmid, 'meshType', meshType, istat)
      call avm_get_realf(avmid, 'modelScale', modelScale, istat)
      call avm_get_real_arrayf(avmid, 'referencePoint', ref_point, 3, istat)

      print '(A,I0,A,A,A,A,A)', "Mesh #", i, " named '", trim(meshName), &
                                "' is type '", trim(meshType), "'"
      print '(A,F5.2)', 'Model Scale is ', modelScale
      print '(A,3F6.2)', 'Reference Point is ', ref_point

      if (meshType=='bfstruc') then
         call avm_get_intf(avmid, 'jmax', jmax, istat)
         call avm_get_intf(avmid, 'kmax', kmax, istat)
         call avm_get_intf(avmid, 'lmax', lmax, istat)
         ! ...
      else if (meshType=='strand') then
         call avm_get_intf(avmid, 'nBndEdges', nBndEdges, istat)
         call avm_get_intf(avmid, 'nPtsPerStrand', nPtsPerStrand, istat)
         ! ...
      else if (meshType=='unstruc') then
         call avm_get_intf(avmid, 'nNodes', nNodes, istat)
         call avm_get_intf(avmid, 'nEdges', nEdges, istat)
         call avm_get_intf(avmid, 'nFaces', nFaces, istat)
         ! ...
      else if (meshType=='cart') then
         call avm_get_intf(avmid, 'nLevels', nLevels, istat)
         call avm_get_intf(avmid, 'nBlocks', nBlocks, istat)
         call avm_get_intf(avmid, 'nFringe', nFringe, istat)
         ! ...
      end if
   end do

   call avm_closef(avmid, istat)
end program example

File: src/rev0/test/example_reader.c
/* 
   This is an example of using the avmesh library with C.
   Error checking is ignored for brevity.
*/

#include <stdio.h>
#include <string.h>
#include "avmesh.h"

int main()
{
   int i, avmid, meshCount;
   double modelScale, ref_point[3];
   char contactName[AVM_STD_STRING_LENGTH];

   int retval = avm_read_headers(&avmid, "mixed.avm");
   if (0 != retval) {
      fprintf(stderr, "ERROR: not an AVMesh file\n");
      return 1;
   }

   avm_get_int(avmid, "meshCount", &meshCount);
   avm_get_real(avmid, "modelScale", &modelScale);
   avm_get_str(avmid, "contactName", contactName, AVM_STD_STRING_LENGTH);
   avm_get_real_array(avmid, "referencePoint", ref_point, 3);

   printf("Mesh Count is %d\n", meshCount);

   for (i=1; i<=meshCount; ++i) {
      char name[AVM_STD_STRING_LENGTH],
           type[AVM_STD_STRING_LENGTH];

      avm_select(avmid, "mesh", i);
      avm_get_str(avmid, "meshName", name, AVM_STD_STRING_LENGTH);
      avm_get_str(avmid, "meshType", type, AVM_STD_STRING_LENGTH);
      avm_get_real(avmid, "modelScale", &modelScale);
      avm_get_real_array(avmid, "referencePoint", ref_point, 3);

      printf("Mesh #%d named '%s' is type '%s'\n", i, name, type);
      printf("Model Scale is %f\n", modelScale);
      printf("Reference Point is %f, %f, %f\n",
             ref_point[0], ref_point[1], ref_point[2]);

      if (0==strcmp(type, "bfstruc")) {
         int jmax,kmax,lmax;
         avm_get_int(avmid, "jmax", &jmax);
         avm_get_int(avmid, "kmax", &kmax);
         avm_get_int(avmid, "lmax", &lmax);
         /* ... */
      }
      else if (0==strcmp(type, "strand")) {
         int nBndEdges,nPtsPerStrand;
         avm_get_int(avmid, "nBndEdges", &nBndEdges);
         avm_get_int(avmid, "nPtsPerStrand", &nPtsPerStrand);
         /* ... */
      }
      else if (0==strcmp(type, "unstruc")) {
         int nNodes,nEdges,nFaces;
         avm_get_int(avmid, "nNodes", &nNodes);
         avm_get_int(avmid, "nEdges", &nEdges);
         avm_get_int(avmid, "nFaces", &nFaces);
         /* ... */
      }
      else if (0==strcmp(type, "cart")) {
         int nLevels,nBlocks,nFringe;
         avm_get_int(avmid, "nLevels", &nLevels);
         avm_get_int(avmid, "nBlocks", &nBlocks);
         avm_get_int(avmid, "nFringe", &nFringe);
         /* ... */
      }
   }

   avm_close(avmid);

   return 0;
}
